[remote tagger] Prevent hanging on shutdown

juliogreff opened the pull request:
### What does this PR do?

In some rare instances where the remote tagger is stopped at the same
time as either the health check or telemetry ticker trigger, the
`select` would not see that `t.Ctx` was already cancelled, and would
carry on trying to re-start the stream. This messes with the shut down
of other agent components, as `Tagger.Stop()` seems to block in that
case.

To make sure that `<-t.ctx.Done()` is always evaluated, we now re-start
the loop on the health check and telemetry ticks.

### Describe how to test/QA your changes

Same as #12592, `error received trying to start stream: tagger stream not started` should not be present 100% of the time (it's been reported that the failure rate was ~30%). Also, this was occasionally breaking kitchen tests. Applying the skip-qa label as this QA is already covered by both of those things.

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

juliogreff committed:
[remote tagger] Don't retry stream when shutting down

We noticed the following error in production, happening multiple times
for the same container:

```
error received trying to start stream: tagger stream not started
```

This happens because the remote tagger tries to reconnect when it should
actually just start shutting down. This commit prevents these retries
from happening, by trying to re-establish a stream at the beginning of a
loop, instead of the end, so it'll only fail once, and not retry.

juliogreff committed:
[remote tagger] Always honor `<-ctx.Done()`

In some rare instances where the remote tagger is stopped at the same
time as either the health check or telemetry ticker trigger, the
`select` would not see that `t.Ctx` was already cancelled, and would
carry on trying to re-start the stream. This messes with the shut down
of other agent components, as `Tagger.Stop()` seems to block in that
case.

To make sure that `<-t.ctx.Done()` is always evaluated, we now re-start
the loop on the health check and telemetry ticks.

vboulineau commented on pkg/tagger/remote/tagger.go:
```diff
@@ -215,8 +215,10 @@ func (t *Tagger) run() {
 	for {
 		select {
 		case <-t.health.C:
+			continue
```
I guess it's on purpose, but the `continue` skips the remaining of the `for` loop entirely, is it expected?

juliogreff replied:
Yes, it's expected. It forces this select to be re-evaluated. If not, a successful `<-t.health.C` will skip a check on `t.ctx.Done`, which may result in a new connection being made instead of an early return.

L3n41c reviewed (commented):
I’m not sure to understand how this change can fix anything.

My understanding of the description of the PR is that there’s an issue if the code after the `for` loop is executed while `t.ctx` has expired.

Adding the `continue` will force to “re-check” if `t.ctx` has expired after the health check or the telemetry ticker has fired.
But `t.ctx` can still expire between the “re-check” that follows the `continue` and the end of the `for { select {` block.

juliogreff commented:
@L3n41c looking at this with fresher eyes today, indeed you're correct. It does reduce the amount of errors (which is what I originally wanted to do with #12592), but would still hang.

There's another check for `ctx.Done` in `startTaggerStream` as a last resort, which errors out early as expected, and it passes the context to the code starting the stream, which I'll assume works as expected. Unfortunately, the error check at the call site is missing another `continue`, so *that's* actually what makes it hang (the code after it is trying to receive on a closed stream), not the expired context itself. As a matter of fact, if we didn't care about the errors none of this context checking is necessary, the missing `continue` does the trick. But like I said, reducing the amount of errors is also desirable, so I'll keep both changes.

juliogreff committed:
[remote tagger] `continue` loop on errors

This prevents a later call to `t.stream.Recv` from hanging on a stopped
stream, if the stream could not be started (in this case, due to the
tagger itself being stopped).

juliogreff commented:
Cherry-picked #12592 (which had been reverted by #12841) in here as well, since it was only reverted in the first place because this fix hadn't been finished timely, and tests were broken.
